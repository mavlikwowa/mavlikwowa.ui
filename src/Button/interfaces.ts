/* Libraries */
import React from 'react';

/* Sizes of button */
export enum ButtonSize {
  SMALL = 'SMALL',
  MEDIUM = 'MEDIUM',
  LARGE = 'LARGE',
}

/* Props of Button component */
export interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
  /* Settings of width (in spite of the size settings, changes the button`s width) */
  width?: string;
  /* A flag of loading (shows a preloader) */
  isLoading?: boolean;
  /* Settings of Background */
  background?: string;
  /* Size */
  buttonSize?: ButtonSize;
  /* Color */
  borderColor?: string;
  /* A showing element */
  // children: React.ReactNode;
}
