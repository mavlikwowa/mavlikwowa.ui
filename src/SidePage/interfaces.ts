/* Libraries */
import React from 'react';

export interface SidePageProps
  extends React.HTMLProps<HTMLDivElement>,
    SidePageContainerProps {
  /* a header`s element */
  header?: React.ReactNode;
  /* An action after click on Close Icon */
  onCloseClick: () => void;
}

export interface SidePageContainerProps {
  /* a showing flag */
  show: boolean;
  /* a color of background */
  background?: string;
}
