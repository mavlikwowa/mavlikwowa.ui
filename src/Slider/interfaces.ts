export interface SliderProps {
  data: React.ReactNode[];
  width?: string;
  height?: string;
}

export interface StyledSliderContentProps {
  slideAmount: number;
  // Important! We should calculate a content width
  width: number;
  height?: string;
}

export interface StyledSliderContainerProps {
  width?: string;
  height?: string;
}
