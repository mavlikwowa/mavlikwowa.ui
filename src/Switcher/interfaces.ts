export interface SwitcherProps extends React.HTMLProps<HTMLSpanElement> {
  hasChildren?: React.ReactNode;
  checked?: boolean;
}

export interface StyledSwitcherProps {
  hasChildren: boolean;
}
