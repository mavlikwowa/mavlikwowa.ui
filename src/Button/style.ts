/* Libraries */
import styled from '@emotion/styled';

/* Interfaces */
import { ButtonProps, ButtonSize } from './interfaces';

/* Styled component */
export const StyledButton = styled.button`
  ${({ width, background, buttonSize }: ButtonProps) => {
    // checks parametrs and creates sizes of button
    const createButtonSize = () => {
      const size = {
        width: 'auto',
        height: '50px',
      };
      if (width) {
        size.width = width;
      } else {
        switch (buttonSize) {
          case ButtonSize.SMALL:
            size.width = '150px';
            size.height = '50px';
            break;
          case ButtonSize.MEDIUM:
            size.width = '200px';
            size.height = '75px';
            break;
          case ButtonSize.LARGE:
            size.width = '250px';
            size.height = '100px';
            break;
          default:
            break;
        }
      }
      return size;
    };
    return `
      display: flex;
      align-items: baseline;
      justify-content: center;
      width: ${createButtonSize().width};
      height: ${createButtonSize().height};
      max-width: 250px;
      background: ${background || 'auto'};
    `;
  }}
`;
