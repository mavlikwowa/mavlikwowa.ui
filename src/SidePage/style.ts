/* Libraries */
import styled from '@emotion/styled';

/* Interfaces */
import { SidePageContainerProps } from './interfaces';

/* Constants */
import { ANIMATION_DURATION } from './const';

/* A styled container */
export const SidePageContainer = styled.div<SidePageContainerProps>`
  ${({ show, background }: SidePageContainerProps) => {
    return `
      display: ${show ? 'block' : 'none'};
      position: fixed;
      top: 0;
      right: 0;
      height: 100%;
      background: ${background || 'white'};
      width: 30%;
      z-index: 999;
      padding: 32px 32px 32px 24px;
      box-sizing: border-box;
      box-shadow: 0 0 10px rgba(0,0,0,0.5);
      transform: translateX(100%);
      transition: transform ${ANIMATION_DURATION}ms;
      
      @media (max-width: 940px) {
        width: 100%;
      }
      &.open {
        transform: translateX(0);
        transition: transform ${ANIMATION_DURATION}ms;
      }
    `;
  }}
`;

/* A styled header */
export const SidePageHeader = styled.div`
  ${() => {
    return `
      display: flex;
      flex-direction: row;
      align-items: center;
      justify-content: space-between;
      margin-bottom: 24px;
      
      > span: last-of-type {
        outline: none;
        cursor: pointer;
        transition: transform 0.5s;
        &:hover {
          transform: rotate(90deg);  
        }  
      }
    `;
  }}
`;
