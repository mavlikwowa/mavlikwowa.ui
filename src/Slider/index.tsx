/* Libraries */
import React, { useState, useEffect, useRef, RefObject } from 'react';

/* Interfaces */
import { SliderProps } from './interfaces';

/* Styled components */
import {
  StyledSlider,
  StyledSliderContent,
  StyledSliderContainer,
  StyledSliderWrapper,
  StyledScrollBar,
} from './style';

/* Components */
import ArrowIcon from '../Icons/ArrowIcon';

/* Hooks */
import { useWindowSize } from '../hooks/useWindowSize';

const Slider: React.FC<SliderProps> = ({
  data,
  width,
  height,
  arrowsColor,
}: SliderProps) => {
  const size = useWindowSize();
  // A sensitivity of swipe
  const SWIPE_AMOUNT = 50;
  // A state of content width. We must calculate it because user can create width in %
  const [cw, setCw] = useState<number>(0);
  // Состояние при начале нажатия на экран
  const [touchStart, setTouchStart] = useState<number>(0);
  // Состояние при завершения нажатия на экран
  const [touchEnd, setTouchEnd] = useState<number>(0);
  // Ref of Content
  const contentRef: RefObject<HTMLDivElement> | null = useRef(null);
  // We should get ref parameters only with a function (nexjs feature)
  const getContentWidth = () => {
    if (contentRef?.current) setCw(contentRef?.current?.offsetWidth);
    return 0;
  };
  // Calculates width after render
  useEffect(() => {
    getContentWidth();
  }, [size, contentRef]);

  // A Number of current slide. Must start from 1 because it`s used in a calculating style
  const [slideAmount, setSlideAmount] = useState<number>(1);
  // Switches to the previous slide
  const prevSlide = () => {
    if (slideAmount >= 1) setSlideAmount(-data.length + 1);
    setSlideAmount((amout) => amout + 1);
  };
  // Switches to the next slide
  const nextSilde = () => {
    if (slideAmount <= -data.length + 2) setSlideAmount(2);
    setSlideAmount((amout) => amout - 1);
  };
  // An event calls after start of swipe
  const handleTouchStart = (e: React.TouchEvent) => {
    setTouchStart(e.targetTouches[0].clientX);
  };
  // An event calls during swipe
  const handleTouchMove = (e: React.TouchEvent) => {
    setTouchEnd(e.targetTouches[0].clientX);
  };
  // Calculates swipe`s results
  const handleTouchEnd = () => {
    if (touchStart - touchEnd > SWIPE_AMOUNT) {
      nextSilde();
    }
    if (touchStart - touchEnd < -SWIPE_AMOUNT) {
      prevSlide();
    }
  };
  return (
    <StyledSliderWrapper width={width}>
      <StyledSlider width={width} height={height} arrowsColor={arrowsColor}>
        <div role="none" data-arrow-left="" onClick={prevSlide}>
          <ArrowIcon />
        </div>
        <StyledSliderContainer
          ref={contentRef}
          width={width}
          height={height}
          onTouchStart={handleTouchStart}
          onTouchMove={handleTouchMove}
          onTouchEnd={handleTouchEnd}
        >
          <StyledSliderContent
            slideAmount={slideAmount}
            width={cw}
            height={height}
          >
            {data?.length > 0 &&
              data.map((item: React.ReactNode, index: number) => {
                // It`s necessary - if I use uuid, the component will update after every touch events
                // eslint-disable-next-line
                return <div key={index}>{item}</div>;
              })}
          </StyledSliderContent>
        </StyledSliderContainer>
        <div role="none" data-arrow-right="" onClick={nextSilde}>
          <ArrowIcon />
        </div>
      </StyledSlider>
      <StyledScrollBar slideAmount={slideAmount} arrowsColor={arrowsColor}>
        {data?.length > 0 && Array(data.length).fill(<div />)}
      </StyledScrollBar>
    </StyledSliderWrapper>
  );
};

export default React.memo(Slider);
