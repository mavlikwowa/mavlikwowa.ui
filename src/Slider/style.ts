/* Libraries */
import styled from '@emotion/styled';

/* Interfaces */
import {
  StyledSliderContentProps,
  SliderProps,
  StyledSliderContainerProps,
} from './interfaces';

export const StyledSlider = styled.div<SliderProps>`
  ${({ width, height }: SliderProps) => {
    return `
      display: flex;
      justify-content: space-between;
      align-items: center;
      width: ${width || '150px'};
      height: ${height || '50px'};
      > div:first-of-type {
        transform: rotate(270deg);
        cursor: pointer;
        outline: none;
      }
      > div:last-of-type {
        transform: rotate(90deg);
        cursor: pointer;
        outline: none;
      }    
    `;
  }}
`;

export const StyledSliderContainer = styled.div<StyledSliderContainerProps>`
  ${({ width, height }: StyledSliderContainerProps) => {
    return `
    display: flex;
    width: ${width || '150px'};
    height: ${height || '50px'};
    justify-content: center;
    align-items: center;
    overflow-x: hidden;
    position: relative;
  `;
  }}
`;

export const StyledSliderContent = styled.div<StyledSliderContentProps>`
  ${({ slideAmount, width, height }: StyledSliderContentProps) => {
    return `
      transform: translateX(${slideAmount * width}px);
      transition: all 1s;
      display: flex;
      justify-content: center;
      align-items: center;
      position: absolute;
      left: -${width}px;
      > div {
        width: ${width}px;
        height: ${height || '50px'};
        display: flex;
        justify-content: center;
        align-items: center;
      }
    `;
  }}
`;
