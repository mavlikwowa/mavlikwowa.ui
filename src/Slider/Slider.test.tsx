/* Libraries */
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

/* Components */
import Slider from './index';
import { StyledSliderContent } from './style';

/* Interfaces */
import { SliderProps } from './interfaces';

const data: React.ReactNode[] = [
  <div>Hi!</div>,
  <div>I`m</div>,
  <div>VOVA</div>,
  <div>!</div>,
];

/** Setups */
const setUp = () => {
  /* A test`s data */
  const props: SliderProps = {
    data,
  };
  /* Component */
  const wrapper = shallow(<Slider {...props} />);
  return { wrapper };
};

/* Tests */
describe('Test component Slider', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    ({ wrapper } = setUp());
  });

  afterEach(() => {
    if (!wrapper) return;
    wrapper.unmount();
  });

  it('should render a component', () => {
    // @todo: key ignore
    expect(wrapper).toMatchSnapshot();
  });

  it('should find a correct text in one child', () => {
    expect(wrapper.find(StyledSliderContent).childAt(0).text()).toBe('Hi!');
  });
});
