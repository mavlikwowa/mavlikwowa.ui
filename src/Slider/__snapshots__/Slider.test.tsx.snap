// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component Slider should render a component 1`] = `ShallowWrapper {}`;
